### PR TITLE
[shardformer] fix master param sync for hybrid plugin/rewrite unwrapping logic

### LINKED ISSUE #4711: [checkpointio] fix master weight sync for hybrid parallel
Checkpoint io object should not be bound to model. We should update hybrid parallel checkpoint io as https://github.com/hpcaitech/ColossalAI/pull/4618

### LINKED ISSUE #4751: [BUG]: self.accelerator.configure(model)
### 🐛 Describe the bug

[`colossalai/booster/booster.py`代码文件第138行](https://github.com/hpcaitech/ColossalAI/blob/main/colossalai/booster/booster.py#L138)，`self.accelerator.configure(model)`

应该改成`self.accelerator.configure_model(model)`

```python
class Accelerator:
    """
    Accelerator is an abstraction for the hardware device that is used to run the model.

    Args:
        device (str): The device to be used. Currently only support 'cpu' and 'gpu'.
    """

    def __init__(self, device: str):
        ...

    def bind(self):
        ...

    def configure_model(self, model: nn.Module) -> nn.Module:
        """
        Move the model to the device.

        Args:
            model (nn.Module): The model to be moved.
        """
        model = model.to(torch.device(self.device))
        return model
```

### Environment

_No response_

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

resolved #4711
resolved #4751 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- rewrite the logic of updating master params in load_sharded_model of HybridPlugin's checkpointIO, disentangle hybrid plugin and its checkpointIO
- rewrite the logic of unwrapping models/optimizers during loading/saving.
- fix the bug mentioned in #4751

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 1]
```diff
@@ -456,42 +457,33 @@ def configure(
...
+                                                        use_pipeline=self.enable_pipeline_parallelism,
+                                                        param_info=param_info,
+                                                        dp_process_group=self.dp_group,
+                                                        tp_process_group=self.tp_group,
+                                                        verbose=True,
+                                                        clip_grad_norm=self.max_norm,
+                                                        **self.zero_config,
+                                                        **self.amp_config)
+        # inject update_master_params
+        model.module.update_master_params = MethodType(optimizer.update_master_params, model)
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:486]: Only inject when optimizer is not None. Maybe you should add a tab

File: colossalai/booster/plugin/low_level_zero_plugin.py [hunk 2]
```diff
@@ -346,7 +317,7 @@ def configure(
                 optimizer, **self.zero_optim_kwargs, verbose=self.verbose
             )
             # inject update_master_params
-            model.update_master_params = MethodType(optimizer.update_master_params, model)
+            model.module.update_master_params = MethodType(optimizer.update_master_params, model)
```
Comment (ver217) [inline colossalai/booster/plugin/low_level_zero_plugin.py:320]: I think we'd better unify the behavior of `unwrap()`. The only effect of `unwrap()` is to get the original model before boosting. General checkpoint io should not unwrap model by default. Whether to unwrap model should be controlled by subclass. And this line should not be changed.

File: colossalai/interface/optimizer.py [hunk 3]
```diff
@@ -133,3 +133,25 @@ def unwrap(self):
         Unwrap the optimizer for checkpoint saving/loading.
         """
         return self.optim
+
+    def update_master_params(self, model: nn.Module) -> None:
```
Comment (ver217) [inline colossalai/interface/optimizer.py:137]: No need to add these methods to optimizer interface.

File: colossalai/booster/plugin/hybrid_parallel_plugin.py (hunk 1 above)
Comment (Fridge003) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py]: modified

File: colossalai/booster/plugin/low_level_zero_plugin.py (hunk 2 above)
Comment (Fridge003) [reply colossalai/booster/plugin/low_level_zero_plugin.py]: modified

File: colossalai/interface/optimizer.py (hunk 3 above)
Comment (Fridge003) [reply colossalai/interface/optimizer.py]: modified

File: colossalai/booster/plugin/gemini_plugin.py [hunk 4]
```diff
@@ -53,24 +54,27 @@ def load_unsharded_model(self, model: GeminiDDP, checkpoint: str, strict: bool =
         Load model from checkpoint with automatic unwrapping.
         The model should be unwrapped in self.load_model via ModelWrapper.unwrap.
         """
-        super().load_unsharded_model(model, checkpoint, strict=strict)
+        assert isinstance(model, GeminiDDP), "Please boost the model before loading!"
+        super().load_unsharded_model(model.unwrap(), checkpoint, strict=strict)
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:58]: I notice that `load_sharded_model` does not unwrap model. Why?

Comment (Fridge003) [reply colossalai/booster/plugin/gemini_plugin.py:58]: sorry, I forgot to unwrap that, now modified.

File: colossalai/booster/plugin/low_level_zero_plugin.py [hunk 5]
```diff
@@ -194,44 +189,23 @@ def load_sharded_optimizer(self, optimizer: OptimizerWrapper, index_file_path: s
...
-        super().save_sharded_model(
-            model.module, checkpoint_path, gather_dtensor, prefix, max_shard_size, use_safetensors
-        )
-
-    def load_unsharded_model(self, model: LowLevelZeroModel, checkpoint: str, strict: bool = True):
-        assert isinstance(model, LowLevelZeroModel)
-        super().load_unsharded_model(model.module, checkpoint, strict)
+    def load_unsharded_model(self, model: ModelWrapper, checkpoint: str, strict: bool = True):
+        assert isinstance(model, LowLevelZeroModel), "Please boost the model before loading!"
+        super().load_unsharded_model(model, checkpoint, strict)
```
Comment (ver217) [inline colossalai/booster/plugin/low_level_zero_plugin.py:196]: Maybe we need to unwrap here. Same as other save/load methods

Comment (Fridge003) [reply colossalai/booster/plugin/low_level_zero_plugin.py:196]: Here the super class is TorchDDPCheckpointIO, unwrapping is done by TorchDDP

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 6]
```diff
@@ -165,6 +166,15 @@ def __init__(self, optim: Optimizer, model: Module, use_pipeline: bool, param_in
             init_pipeline_optimizer(optim, model)
         super().__init__(optim)
 
+    def update_master_params(self, model: Module):
+        return None
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:170]: `pass` should be clearer if it's just a placeholder here.

**Review by ver217 — APPROVED** [review]

**Review by CjhHa1 — APPROVED** [review]
LGTM

**Pull request merged**